### PR TITLE
Introduce metrics, a metric event library.

### PR DESCRIPTION
**Pull request by stevegury**
The library is composed of two central objects: the `Recorder` and the `Aggregator`.

You use the `Recorder` to create `Counter` and `Timer`, then the `Counter` and
`Timer` are used to send events to the `Recorder`.

On the other end, the `Aggregator` listen to events emitted by the `Counter` / `Timer`
and aggregate them. The default `Aggregator` aggregates counter events as a single
value (the latest) and timer event into a histogram.
But nothing is preventing you from having different aggregating strategy, for
instance you may want to keep a min/avg/latest/max structure for every counter.

The library provides two histograms implementations, one based on the classical
bucketed algorithm and one based on the more innovative frugal streaming
estimation. The first one is as precise as you want with an impact on speed and
memory footprint, the second one is extremely fast with almost no memory impact
but is less precise.

Here's the performance benchmark on my laptop:
```
$ time bin/metricsBench.js
{
  "iterations": 200000,
  "warmUp": 3,
  "tries": 10,
  "unit": "µs",
  "create counter": {
    "min": "0.125",
    "avg": "0.134",
    "max": "0.155"
  },
  "increment counter": {
    "min": "0.045",
    "avg": "0.047",
    "max": "0.060"
  },
  "create disabled counter": {
    "min": "0.030",
    "avg": "0.032",
    "max": "0.030"
  },
  "increment disabled counter": {
    "min": "0.010",
    "avg": "0.013",
    "max": "0.015"
  },
  "create timer (bucket)": {
    "min": "0.070",
    "avg": "0.072",
    "max": "0.080"
  },
  "start/stop timer (bucket)": {
    "min": "0.340",
    "avg": "0.365",
    "max": "0.405"
  },
  "create tag timer (bucket)": {
    "min": "0.075",
    "avg": "0.077",
    "max": "0.075"
  },
  "start/stop tag timer (bucket)": {
    "min": "2.860",
    "avg": "2.928",
    "max": "2.985"
  },
  "create timer (streaming)": {
    "min": "0.090",
    "avg": "0.095",
    "max": "0.110"
  },
  "start/stop timer (streaming)": {
    "min": "0.225",
    "avg": "0.236",
    "max": "0.295"
  },
  "create disabled timer": {
    "min": "0.015",
    "avg": "0.017",
    "max": "0.015"
  },
  "start/stop disabled timer": {
    "min": "0.020",
    "avg": "0.019",
    "max": "0.020"
  },
  "aggregator report (bucket histo)": {
    "min": "394.000",
    "avg": "402.898",
    "max": "408.995"
  },
  "aggregator report (streaming histo)": {
    "min": "200.000",
    "avg": "211.099",
    "max": "225.995"
  },
  "disabled-aggregator report": {
    "min": "0.000",
    "avg": "0.199",
    "max": "0.995"
  }
}
```

### CODE REVIEW

File: bin/metricsBench.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env node
+'use strict';
+
+var _ = require('lodash');
+
+var getSemaphore = require('../test/common/getSemaphore.js');
```
Comment (yunong) [inline bin/metricsBench.js]: This is nitty and style related, but we tend to organise imports in the C style, which means there are 3 sections, 1) core modules, 2) npm modules, 3) internal files. Each section is then alphabetically sorted.

**Comment by yunong** [comment]
Just a general comment, for `metricsBench` can we output in JSON? That'll make it easy to consume results.

File: lib/metrics/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# metrics
+
+## Architecture
+
+The library is composed of two central objects: the `Recorder` and the `Aggregator`.
+
+You use the `Recorder` to create `Counter`s and `Timer`s, then the `Counter`s and
```
Comment (yunong) [inline lib/metrics/README.md]: should be `Counters` and `Timers`, otherwise the formatting looks off

File: lib/metrics/aggregator.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * The aggregator listen to the aggregator events (counter/timer) and aggregates
+ * those events into respectively values/histograms.
+ *
+ * This abstraction reflect the "separation of concern" strategy used in this
+ * library, the aggregator decides how it want to aggregate the values. It may
+ * be interested in specific metrics and decide to aggregate them with an higher
+ * precision.
+ *
+ * @param {Object} recorder Recorder the aggregator listen to.
+ * @param {Object} _config Config object.
```
Comment (yunong) [inline lib/metrics/aggregator.js]: why is this `_config` and not just `config`?

File: lib/metrics/aggregator.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ * @param {Func} [_config.histogram] the function used to create an histogram.
```
Comment (yunong) [inline lib/metrics/aggregator.js]: s/Func/Function

File: lib/metrics/aggregator.js (hunk 3 above)
Comment (stevegury) [inline lib/metrics/aggregator.js]: I defined `config` as `var config = _config || DefaultConfig.aggregator;`
This simplified the logic.

File: bin/metricsBench.js [hunk 5]
```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env node
+'use strict';
+
+var _ = require('lodash');
+
+var Aggregator = require('../lib/metrics/aggregator.js');
+var BucketedHistogram = require('../lib/metrics/stats/bucketedhistogram.js');
+var getRandomInt = require('../test/common/getRandomInt.js');
```
Comment (yunong) [inline bin/metricsBench.js]: The `test` directory generally isn't included when you publish an npm module. If we want this file shipped as part of the distribution, we should consider moving these utils into `lib`.

**Comment by stevegury** [comment]
Let's close that, I moved the metric code to [its own repo](github.com/stevegury/metrix).

**Pull request closed without merging**